### ISSUE TITLE
Memory leak / performance issue

### ISSUE THREAD

**Issue by Harrisonl** [guidesmiths/rascal#161]
Hi 👋  loving rascal so far!

However, we are running into an issue in production. Basically, we have a somewhat substantial write throughput (well not that large probably around 1k/sec). 

The issue we are seeing is, if we don't await for the `publication.on('success', resolve)` callback, we will see highly degraded publication to rabbitMq (like 60/s) and growing memory for our instance (typically around 5% - started growing past 40%). This is also with like 0/s reads, where it should be quite high (e.g. reads are just not happening).

It appears the messages are getting stuck on the server and not actually being published to rabbitmq unless we use the following piece of code and everywhere we publish a message, however - from the docs this doesn't seem like it's necessary. We can verify this because once we kill the process that writes the messages, we still see messages being published for a period of time.

```ts
const pub = (......) => {
    const pubber = await broker.publish(getExchange(event), JSON.stringify(payload), {
      routingKey: event,
      options: { contentEncoding: "application/json", messageId: v4() },
    })

    return new Promise((resolve, reject) => {
      pubber.on("error", (err) => {
        reject(err)
      })
      pubber.on("success", resolve)
    })
  }

``` 

Even when this is there, the reads are quite slow - (roughly 84/second - which is actually slower then when we had more work being done in each consumption, partly I'm assuming because of the related issue because now It writes to a queue instead of doing some of the work) where we'd expect it to be much higher.

Pool wise we have it set to min: 5, max: 10. For reads we have 3 consumers, 10 prefetch.

Rascal version: `13.0.3` 

Basically - would love to know if there's any gotchas that we are missing or any obvious causes for the slowdown in writes and reads as well as the apparent bottleneck / memory leak with the case of no awaiting success.

**Comment by Harrisonl** [comment]
Another interesting thing I've noticed is that our deliver/get will almost always be identical p/s for every queue, regardless the number of channels and prefetch...which seems suspicious that we are doing something wrong because I wouldn't expect there such a consistent throughput for 3 completely separate operations. 

![Screenshot 2021-07-18 at 18 50 24](https://user-images.githubusercontent.com/10399120/126077236-e5ac8836-37f5-4025-8afd-82ad41714867.png)

**Comment by cressie176** [comment]
Hi @Harrisonl,

I suspect the root cause is that you are producing messages faster that Rascal can publish them. Consequently they're building up in the event loop, and consuming a lot of memory, and thrashing the application due to garbage collection. This isn't a leak, but has similar side effects. The only ways to fix are 

1. Find a way to publish messages more quickly
2. Find a way to throttle production of the messages, so they don't accumulate

#### Publish messages more quickly
The most effective way to speed up publishing is to use regular channels rather than confirm channels by specifying `confirm: false` in the publication config. The price you pay is that they're fire and forget, so you have no assurance that your message reached the broker. Given that you weren't originally waiting for the promise to resolve, you might be happy with this compromise though.

If you need confirm channels, then you can try tuning the pool settings, but even then you may find Rascal too slow. The underlying issue is because of a design decision not to share channels between publishes. Also there's only one connection to a vhost, so this can get saturated too. It's rather hacky but you can work around this by creating multiple vhost configurations with the same details and cycling between them.

#### Throttle Production
Rascal will emit a "busy" event when all the channels in the pool are saturated. You may be able to use this to throttle whatever is producing the messages. For example, if you are streaming messages from a mongo collection, you should be able to apply back pressure to the stream. Alternatively if the source of the messages are HTTP requests, you may consider returning a 503.
When one of the channel's clears, Rascal emits a "ready" event, which you can use to release the back pressure. Again deatils are in the [README](https://github.com/guidesmiths/rascal#flow-control)

#### Slow consumption
I suspect this is because when you wait for the promise to resolve, the messages aren't published fast enough, and you're starving the consumers. Alternatively it could be because you are running the consumers in the same environment as the publisher and the CPU or Memory as being hogged.

#### Other Stuff
There should no need to stringify the payload. Rascal does this for you using a safe stringifier. It also defaults to "application/json" and will generate a uuid.

**Comment by Harrisonl** [comment]
@cressie176 Thanks so much for the detailed reply/suggestions!

We do have confirm channels off - but will look into your other suggestions of vhost cycling - just to confirm there is no way to open more then one connection at once in a single vhost config?

We are running the consumers in the same environment as the publishers (which is probably further impacting the problem). We are making a change to split this out and will give you an update if this helps.

**Comment by cressie176** [comment]
>  just to confirm there is no way to open more than one connection at once in a single vhost config?

No, unfortunately not. I'd like to add it, but the VHost channel pooling is already difficult enough to manage, so I'm reluctant to add a connection pool as well.

**Comment by cressie176** [comment]
Any luck @Harrisonl?
